### PR TITLE
Editing: auto closing brace, bracket and parenthesis

### LINKED ISSUE #23: Editing: auto closing brace, bracket and parenthesis
# What would you like to be added
When the user types a  brace, bracket or parenthesis in a rego, it should be automatically close by IDE and caret should be placed between the braces, brackets, and parenthesis

# Why is this needed
For better user experience when editing rego file.

# Implementation details

We have to implement a `PairedBraceMatcher`.

An example are available [here](https://github.com/intellij-rust/intellij-rust/blob/6c20feb18c7983795eb7c6f951e8878ee0d42b1c/intellij-toml/src/main/kotlin/org/toml/ide/TomlBraceMatcher.kt#L12) and [here](https://github.com/intellij-rust/intellij-rust/blob/b9fdd972c64aafb91266f1c90401a2834d2b1f5d/src/main/kotlin/org/rust/ide/typing/RsBraceMatcher.kt#L130-L129)

### PR DESCRIPTION
**Pull request by vgramer**
# Description
Implements a `PairedBraceMatcher` to auto close brace, parenthesis and bracket 
 

Fixes #23 

# Special notes for your reviewer

### CODE REVIEW

**Comment by vgramer** [comment]
I change this PR to draft the time to investigate why tests are falling in GitHub action and not on my laptop

File: src/main/kotlin/org/openpolicyagent/ideaplugin/ide/typing/RegoBraceMatcher.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    override fun isPairedBracesAllowedBeforeType(lbraceType: IElementType, contextType: IElementType?): Boolean = true
+
+    override fun getCodeConstructStart(file: PsiFile, openingBraceOffset: Int): Int = openingBraceOffset
+
+    companion object {
+        private val PAIRS: Array<BracePair> = arrayOf(
+            BracePair(RegoTypes.LBRACE, RegoTypes.RBRACE, true),
+            BracePair(RegoTypes.LBRACK, RegoTypes.RBRACK, false),
+            BracePair(RegoTypes.LPAREN, RegoTypes.RPAREN, false)
```
Comment (asadali) [inline src/main/kotlin/org/openpolicyagent/ideaplugin/ide/typing/RegoBraceMatcher.kt:29]: :+1:

File: src/test/kotlin/org/openpolicyagent/ideaplugin/ide/typing/RegoBraceMatcherTest.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        myFixture.configureByText(RegoFileType, before)
+        myFixture.type(type)
+        myFixture.checkResult(after)
+    }
+
+    private fun doMatch(source: String, coBrace: String) {
+        myFixture.configureByText(RegoFileType, source)
+        val expected = source.replace("<caret>", "").lastIndexOf(coBrace)
+        check(BraceMatchingUtil.getMatchedBraceOffset(myFixture.editor, true, myFixture.file) == expected)
+    }
```
Comment (asadali) [inline src/test/kotlin/org/openpolicyagent/ideaplugin/ide/typing/RegoBraceMatcherTest.kt:122]: :+1: these test cases are concise and elegant! :)

**Review by asadali — APPROVED** [review]

File: src/test/kotlin/org/openpolicyagent/ideaplugin/ide/typing/RegoBraceMatcherTest.kt (hunk 2 above)
Comment (vgramer) [reply src/test/kotlin/org/openpolicyagent/ideaplugin/ide/typing/RegoBraceMatcherTest.kt:122]: thanks but it's not my idea, it Jetbrains team ;)

**Comment by vgramer** [comment]
> I change this PR to draft the time to investigate why tests are falling in GitHub action and not on my laptop

Test pass again, I think it was the bug about `core.xml` plugin descriptor

**Pull request merged**